File: rb/test/incident_entity_test.rb
# Incident entity test

require "minitest/autorun"
require "json"
require_relative "../Statuspage_sdk"
require_relative "runner"

class IncidentEntityTest < Minitest::Test
  def test_create_instance
    testsdk = StatuspageSDK.test(nil, nil)
    ent = testsdk.Incident(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = incident_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "update", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "incident." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set STATUSPAGE_TEST_INCIDENT_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    incident_ref01_ent = client.Incident(nil)
    incident_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.incident"), "incident_ref01"))
    incident_ref01_data["page_id"] = setup[:idmap]["page01"]

    incident_ref01_data_result = incident_ref01_ent.create(incident_ref01_data, nil)
    incident_ref01_data = Helpers.to_map(incident_ref01_data_result)
    assert !incident_ref01_data.nil?
    assert !incident_ref01_data["id"].nil?

    # LIST
    incident_ref01_match = {
      "page_id" => setup[:idmap]["page01"],
    }

    incident_ref01_list_result = incident_ref01_ent.list(incident_ref01_match, nil)
    assert incident_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(incident_ref01_list_result),
      { "id" => incident_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # UPDATE
    incident_ref01_data_up0_up = {
      "id" => incident_ref01_data["id"],
      "page_id" => setup[:idmap]["page_id"],
    }

    incident_ref01_markdef_up0_name = "created_at"
    incident_ref01_markdef_up0_value = "Mark01-incident_ref01_#{setup[:now]}"
    incident_ref01_data_up0_up[incident_ref01_markdef_up0_name] = incident_ref01_markdef_up0_value

    incident_ref01_resdata_up0_result = incident_ref01_ent.update(incident_ref01_data_up0_up, nil)
    incident_ref01_resdata_up0 = Helpers.to_map(incident_ref01_resdata_up0_result)
    assert !incident_ref01_resdata_up0.nil?
    assert_equal incident_ref01_resdata_up0["id"], incident_ref01_data_up0_up["id"]
    assert_equal incident_ref01_resdata_up0[incident_ref01_markdef_up0_name], incident_ref01_markdef_up0_value

    # LOAD
    incident_ref01_match_dt0 = {
      "id" => incident_ref01_data["id"],
    }
    incident_ref01_data_dt0_loaded = incident_ref01_ent.load(incident_ref01_match_dt0, nil)
    incident_ref01_data_dt0_load_result = Helpers.to_map(incident_ref01_data_dt0_loaded)
    assert !incident_ref01_data_dt0_load_result.nil?
    assert_equal incident_ref01_data_dt0_load_result["id"], incident_ref01_data["id"]

    # REMOVE
    incident_ref01_match_rm0 = {
      "id" => incident_ref01_data["id"],
    }
    incident_ref01_ent.remove(incident_ref01_match_rm0, nil)

    # LIST
    incident_ref01_match_rt0 = {
      "page_id" => setup[:idmap]["page01"],
    }

    incident_ref01_list_rt0_result = incident_ref01_ent.list(incident_ref01_match_rt0, nil)
    assert incident_ref01_list_rt0_result.is_a?(Array)

    not_found_item = Vs.select(
      Runner.entity_list_to_data(incident_ref01_list_rt0_result),
      { "id" => incident_ref01_data["id"] })
    assert Vs.isempty(not_found_item)

  end
end

def incident_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "incident", "IncidentTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = StatuspageSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["incident01", "incident02", "incident03", "page01", "page02", "page03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["STATUSPAGE_TEST_INCIDENT_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "STATUSPAGE_TEST_INCIDENT_ENTID" => idmap,
    "STATUSPAGE_TEST_LIVE" => "FALSE",
    "STATUSPAGE_TEST_EXPLAIN" => "FALSE",
    "STATUSPAGE_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["STATUSPAGE_TEST_INCIDENT_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end
  if idmap_resolved["page_id"].nil?
    idmap_resolved["page_id"] = idmap_resolved["page01"]
  end

  if env["STATUSPAGE_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["STATUSPAGE_APIKEY"],
      },
      extra || {},
    ])
    client = StatuspageSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["STATUSPAGE_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["STATUSPAGE_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/incident_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/statuspage-sdk/go"
	"github.com/voxgig-sdk/statuspage-sdk/go/core"

	vs "github.com/voxgig-sdk/statuspage-sdk/go/utility/struct"
)

func TestIncidentEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Incident(nil)
		if ent == nil {
			t.Fatal("expected non-nil IncidentEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := incidentBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "incident." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set STATUSPAGE_TEST_INCIDENT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		incidentRef01Ent := client.Incident(nil)
		incidentRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "incident"}, setup.data), "incident_ref01"))
		incidentRef01Data["page_id"] = setup.idmap["page01"]

		incidentRef01DataResult, err := incidentRef01Ent.Create(incidentRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		incidentRef01Data = core.ToMapAny(incidentRef01DataResult)
		if incidentRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if incidentRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		incidentRef01Match := map[string]any{
			"page_id": setup.idmap["page01"],
		}

		incidentRef01ListResult, err := incidentRef01Ent.List(incidentRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		incidentRef01List, incidentRef01ListOk := incidentRef01ListResult.([]any)
		if !incidentRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", incidentRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(incidentRef01List), map[string]any{"id": incidentRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		incidentRef01DataUp0Up := map[string]any{
			"id": incidentRef01Data["id"],
			"page_id": setup.idmap["page_id"],
		}

		incidentRef01MarkdefUp0Name := "created_at"
		incidentRef01MarkdefUp0Value := fmt.Sprintf("Mark01-incident_ref01_%d", setup.now)
		incidentRef01DataUp0Up[incidentRef01MarkdefUp0Name] = incidentRef01MarkdefUp0Value

		incidentRef01ResdataUp0Result, err := incidentRef01Ent.Update(incidentRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		incidentRef01ResdataUp0 := core.ToMapAny(incidentRef01ResdataUp0Result)
		if incidentRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if incidentRef01ResdataUp0["id"] != incidentRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if incidentRef01ResdataUp0[incidentRef01MarkdefUp0Name] != incidentRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", incidentRef01MarkdefUp0Name, incidentRef01ResdataUp0[incidentRef01MarkdefUp0Name])
		}

		// LOAD
		incidentRef01MatchDt0 := map[string]any{
			"id": incidentRef01Data["id"],
		}
		incidentRef01DataDt0Loaded, err := incidentRef01Ent.Load(incidentRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		incidentRef01DataDt0LoadResult := core.ToMapAny(incidentRef01DataDt0Loaded)
		if incidentRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if incidentRef01DataDt0LoadResult["id"] != incidentRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		incidentRef01MatchRm0 := map[string]any{
			"id": incidentRef01Data["id"],
		}
		_, err = incidentRef01Ent.Remove(incidentRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		incidentRef01MatchRt0 := map[string]any{
			"page_id": setup.idmap["page01"],
		}

		incidentRef01ListRt0Result, err := incidentRef01Ent.List(incidentRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		incidentRef01ListRt0, incidentRef01ListRt0Ok := incidentRef01ListRt0Result.([]any)
		if !incidentRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", incidentRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(incidentRef01ListRt0), map[string]any{"id": incidentRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func incidentBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "incident", "IncidentTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read incident test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse incident test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"incident01", "incident02", "incident03", "page01", "page02", "page03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("STATUSPAGE_TEST_INCIDENT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"STATUSPAGE_TEST_INCIDENT_ENTID": idmap,
		"STATUSPAGE_TEST_LIVE":      "FALSE",
		"STATUSPAGE_TEST_EXPLAIN":   "FALSE",
		"STATUSPAGE_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["STATUSPAGE_TEST_INCIDENT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add page_id alias for update test.
	if idmapResolved["page_id"] == nil {
		idmapResolved["page_id"] = idmapResolved["page01"]
	}

	if env["STATUSPAGE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["STATUSPAGE_APIKEY"],
			},
			extra,
		})
		client = sdk.NewStatuspageSDK(core.ToMapAny(mergedOpts))
	}

	live := env["STATUSPAGE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["STATUSPAGE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
